src/main/java/org/myrobotlab/service/Mqtt.java

```diff
@@ -1,439 +1,538 @@
+
+			// what if you don't want to do x-forwarding ???
+			inbox.add(msg);
+
+			// COMMON GATEWAY REGISTERATION AND X-FORWARDED END ----------------
+
+		} catch (Exception e) {
+			error("tried decoding [%] into a message", payload);
+		}
+
+		// are all necessary ?
+		invoke("publishMqttMsg", topic, message.getPayload());
+		invoke("publishMqttMsgByte", message.getPayload());
+		invoke("publishMqttMsgString", new String(message.getPayload()), topic);
+	}
+
+	@Override
+	public void onFailure(IMqttToken token, Throwable throwable) {
+		log.error("error - {} {}", tokenToString(token), throwable);
+	}
+
+	@Override
+	public void onSuccess(IMqttToken token) {
+		log.info("success - {} ", tokenToString(token));
+	}
+
+	// dangerous - as topic is a field value
+	public void publish(String msg) throws Throwable {
+		publish(topic, qos, msg);
+	}
+
+	public void publish(String topic, Integer qos, byte[] payload) throws MqttPersistenceException, MqttException {
+		MqttMessage message = null;
+
+		if (client == null || !client.isConnected()) {
+			connect(url);
+		}
+
+		// Send / publish a message to the server
+		// Get a token and setup an asynchronous listener on the token which
+		// will be notified once the message has been delivered
+		message = new MqttMessage(payload);
+		message.setQos(qos);
+
+		String time = new Timestamp(System.currentTimeMillis()).toString();
+		log.info("Publishing at: " + time + " to topic \"" + topic + "\" qos " + qos);
+
+		// FIXME - see if user context is like a backend id ...
+		client.publish(topic, message, "Pub sample context", this);
+
+	}
+
+	public void publish(String topicName, int qos, String payload) throws MqttPersistenceException, MqttException {
+		publish(topicName, qos, payload.getBytes());
+	}
+
+	public MqttMsg publishMqttMsg(String topic, byte[] msg) {
+		return new MqttMsg(topic, msg);
+	}
+
+	public String[] publishMqttMsgString(String msg, String topic) {
+		String[] result = { msg, topic };
+		return result;
+	}
+
+	public byte[] publishMqttMsgByte(byte[] msg) {
+		return msg;
+	}
+
+	public void setBroker(String broker) {
+		url = broker;
+	}
+
+	public void setClientId(String cId) {
+		clientId = cId;
+	}
+
+	public void setPubTopic(String Topic) {
+		topic = Topic;
+	}
+
+	public void setQos(int q) {
+		qos = q;
+	}
+
+	public void subscribe(String topic) throws MqttException {
+		subscribe(topic, 2);
+	}
+
+	/**
+	 * Subscribe to a topic on an Mqtt server Once subscribed this method waits for
+	 * the messages to arrive from the server that match the subscription. It
+	 * continues listening for messages until the enter key is pressed. {
+	 * 
+	 * @param topic
+	 *            to subscribe to (can be wild carded)
+	 * 
+	 * @param qos
+	 *            the maximum quality of service to receive messages at for this
+	 *            subscription
+	 */
+	public void subscribe(String topic, int qos) throws MqttException {
+		client.subscribe(topic, qos, "Subscribe sample context", this);
```

Comment:
`"Subscribe sample context"` seems like a weird hardcoded value.

Reply:
agreed, but at the moment do not know the use of userContext